pandas/_libs/src/inference.pyx

```diff
@@ -606,9 +611,16 @@ cpdef bint is_bool_array(ndarray values):
         if n == 0:
             return False
 
-        for i in range(n):
-            if not util.is_bool_object(objbuf[i]):
-                return False
+        if skipna:
+            for i in range(n):
+                val = objbuf[i]
+                if not util._checknull(val) and not util.is_bool_object(val):
+                    return False
```

Comment:
I'm not sure I follow you here.  This loop would only be entered if `skipna` is `True`.  What I meant is that you could potentially abstract this for-loop into a separate function, which accepts `ndarray`, `n`, and a `dtype`-checker function (e.g. `is_bool_object`, `is_float_object`).

Would be more compact but could impact performance, which is why I'm not pushing this so hard but just floating the idea.